Platform/Authentication/Token Provider/DefaultTokenProvider.swift

```diff
@@ -0,0 +1,228 @@
+import Foundation
+
+/// DefaultTokenProvider makes calls to a specified HTTPS endpoint and expects to receive a OAuth 2.0
+/// token from it.
+///
+/// This is the implementation we recommend for production use, to request tokens from your backend
+/// system.
+///
+/// If this class does not fit your needs, you can implement the `TokenProvider` protocol yourself.
+public class DefaultTokenProvider: TokenProvider {
+    
+    // MARK: - Types
+    
+    /// The closure used by the token provider in order to obtain dictionary of headers.
+    public typealias HeadersInjector = () -> [String : String]?
+    
+    /// The closure used by the token provider in order to obtain list of query items.
+    public typealias QueryItemsInjector = () -> [URLQueryItem]?
+    
+    /// The closure used by the token provider in order to obtain list of URL encoded body items.
+    public typealias BodyInjector = () -> [URLEncodedBodyItem]?
+    
+    // MARK: - Properties
+    
+    /// The URL that will be used by the token provider to retrieve a token.
+    public let url: URL
+    
+    /// An optional dictionary of headers to include in the request. Here you can supply the session
+    /// or other credientials which your endpoint might require to authenticate the request.
+    ///
+    /// If not specified otherwise, the token provider will always set Content-Type header
+    /// to application/x-www-form-urlencoded.
+    public private(set) var headers: [String : String]?
+    
+    /// An optional list of query items to include in the request URL. Here you can supply any query items
+    /// which your endpoint might require to process the request.
+    public private(set) var queryItems: [URLQueryItem]?
+    
+    /// An optional list of URL encoded body items to include in the request. Here you can supply any
+    /// body items which your endpoint might require to process the request.
+    ///
+    /// If not specified otherwise, the token provider will always add grant_type=client_credentials item
+    /// to the body of the request.
```

Comment:
This comment about "grant_type" appears in a few places but I don't actually see this behaviour implemented anywhere?

Reply:
This is implemented in the following method of `DefaultTokenProvider` and tested in Functional Tests target:

```swift
private func requestBody(for body: [URLEncodedBodyItem]?) -> [URLEncodedBodyItem]
```